src/box/box.cc

```diff
@@ -681,6 +681,141 @@ box_check_uri(const char *source, const char *option_name)
+		if (source[0] == delimiters[i]) {
+			diag_set(ClientError, ER_CFG, "listen", error);
+			return -1;
+		}
+	}
+	return 0;
+}
+
+static inline int
+box_check_end_of_token(const char *delimiter, char **saveptr,
+		       const char *error)
+{
+	if (strtok_r(NULL, delimiter, saveptr) != NULL) {
+		diag_set(ClientError, ER_CFG, "listen", error);
+		return -1;
+	}
+	return 0;
+}
+
+static inline int
+box_check_uri_option(const char *opt, const char *val)
+{
+	if (!strcmp(opt, "transport")) {
+		if (strcmp(val, "plain")) {
+			diag_set(ClientError, ER_CFG, "listen",
+				 tt_sprintf("invalid value `%s` for "
+				 	    "`transport` option", val));
+			return -1;
+		}
+	} else {
+		diag_set(ClientError, ER_CFG, "listen",
+			 tt_sprintf("invalid option `%s` for URI", opt));
+		return -1;
+	}
+	return 0;
+}
+
+static int
+box_check_uri_options(char *opts)
+{
+	static const char *error = "unable to parse URI options, extra `=`";
+	char *saveptr1, *saveptr2, *saveptr3;
+	for (char *optval = opts; opts != NULL; optval = NULL) {
+		char *optval_token = strtok_r(optval, "&", &saveptr1);
+		if (optval_token == NULL)
+                	break;
+                char *opt = strtok_r(optval_token, "=", &saveptr2);
+                char *vals = strtok_r(NULL, "=", &saveptr2);
+                if (opt == NULL) {
+                	diag_set(ClientError, ER_CFG, "listen",
+                		 "invalid value `null` for URI option");
+                	return -1;
+                }
+                if (vals == NULL) {
+                	diag_set(ClientError, ER_CFG, "listen",
+                		 tt_sprintf("invalid value `null` for URI "
+                		 	    "`%s` option", opt));
+                	return -1;
+                }
+                if (box_check_end_of_token("=", &saveptr2, error) != 0)
+			return -1;
+                for (char *val = vals; vals != NULL; val = NULL) {
+                	char *val_token = strtok_r(val, ";", &saveptr3);
+                	if (val_token == NULL)
+                		break;
+                	if (box_check_uri_option(opt, val_token) != 0)
+                		return -1;
+                }
+	}
+	return 0;
+}
+
+static int
+box_check_uri_with_options(const char *source)
+{
+	int rc = 0;
+	char *uri_with_opts, *uri_with_opts_token;
+	char *saveptr1, *saveptr2;
+	static const char *uri_error = "expected host:service or /unix.socket";
+	static const char *uri_opt_error = "unable to parse URI options, extra `?`";
+
+	char *source_copy = strdup(source);
+	if (source_copy == NULL) {
+		diag_set(OutOfMemory, strlen(source) + 1,
+			 "strdup", "source");
+		return -1;
+	}
+
+	if (box_check_delimiters(source_copy, ", \0", 3, uri_error) != 0) {
+		rc = -1;
+		goto cleanup;
+	}
+
+	for (uri_with_opts = source_copy; ; uri_with_opts = NULL) {
+		uri_with_opts_token =
+			strtok_r(uri_with_opts, ", ", &saveptr1);
+		if (uri_with_opts_token == NULL)
+			break;
+		rc = box_check_delimiters(uri_with_opts_token, "?", 1, uri_error);
```

Comment:
You pass in fact the same error, mb it is worth incapsulating it in box_check_delimiters

Reply:
fixed